File: leetcode_py/Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def addTwoNumbers(self, l1, l2):
        dummy=ListNode(-1)
        head=dummy
        carry=0

        while l1 or l2:
            tmpres=0
            if l1:
                tmpres+=l1.val
                l1=l1.next
            if l2:
                tmpres+=l2.val
                l2=l2.next
            tmpres+=carry
            carry=tmpres/10
            tmpres%=10
            head.next=ListNode(tmpres)
            head=head.next
        if carry:
            head.next=ListNode(carry)
        return dummy.next
    # Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def moretoend(self,new,up):
        while new.next is not None:
            new.val=new.val+up
            up=new.val/10
            new.val%=10
            new=new.next

        new.val=new.val+up
        up=new.val/10
        new.val%=10
        if up>0:
            new.next=ListNode(up)


    # @return a ListNode
    def addTwoNumbers(self, l1, l2):
        up=0
        if l1==None:
            return l2
        if l2==None:
            return l1
        head=l1
        while l1.next is not None and l2.next is not None:
            l1.val=l1.val+l2.val+up
            up=l1.val/10
            l1.val%=10
            l1=l1.next
            l2=l2.next

        if l2.next is None and l1.next is None:
            l1.val+=l2.val+up
            up=l1.val/10
            l1.val%=10
            if up>0:
                l1.next=ListNode(up)
        elif l2.next is None:
            l1.val+=l2.val
            self.moretoend(l1,up)
        elif l1.next is None:
            l1.val+=l2.val
            l1.next=l2.next
            self.moretoend(l1,up)

        return head

File: leetcode_java/add-two-numbers/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode p = l1 , q = l2 , temp = null , pre = null;
        int c = 0;
        int sum ;
        while(p != null && q != null){
            sum = p.val + q.val + c;
            c =  sum / 10;
            p.val = sum % 10;
            pre = p;
            p = p.next ;
            q = q.next;
        }
        if(p == null){
            pre.next =  q ;
        }
        temp = pre.next ;
        if(temp == null && c > 0){
             ListNode newNode = new ListNode(c);
             newNode.next = null;
            pre.next =  newNode;
        }
        while(temp != null && temp.next != null){
            temp.val = temp.val + c;
            c= temp.val / 10;
            temp.val = temp.val % 10;
            temp = temp.next;
        }
        int val = 0;
        if(temp != null){
            val = temp.val+c;
            temp.val =  val % 10 ;
            c= val / 10;
            
           
        }
        if(temp != null && c >0 ){
            ListNode node = new ListNode(c );
            node.next = null;
            temp.next = node;
            
        }
        
        return l1;
       
    }